ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto

```diff
@@ -12,30 +12,63 @@ option java_outer_classname = "TransactionFilterOuterClass";
 option java_package = "com.daml.ledger.api.v1";
 option csharp_namespace = "Com.Daml.Ledger.Api.V1";
 
-// Used for filtering Transaction and Active Contract Set streams.
-// Determines which on-ledger events will be served to the client.
+// A filter both for filtering create and archive events as well as for
+// filtering transaction trees.
 message TransactionFilter {
 
-  // Keys of the map determine which parties' on-ledger transactions are being queried.
-  // Values of the map determine which events are disclosed in the stream per party.
-  // At the minimum, a party needs to set an empty Filters message to receive any events.
   // Each key must be a valid PartyIdString (as described in ``value.proto``).
+  // The interpretation of the filter depends on the stream being filtered:
+  // (1) For **transaction tree streams** only party filters with wildcards are allowed, and all subtrees
+  //     whose root has one of the listed parties as an informee are returned.
+  // (2) For **transaction and active-contract-set streams** create and archive events are returned for all contracts whose
+  //    stakeholders include at least one of the listed parties and match the
+  //    per-party filter.
   // Required
   map<string, Filters> filters_by_party = 1;
 }
 
+// The union of a set of contract filters, or a wildcard.
 message Filters {
 
-  // If not set, no filters will be applied.
+  // If set, then contracts matching any of the ``InclusiveFilters`` match
+  // this filter.
+  // If not set, any contract matches this filter.
   // Optional
   InclusiveFilters inclusive = 1;
 }
 
-// If no internal fields are set, no filters will be applied.
+// A filter that matches all contracts that are either an instance of one of
+// the ``template_ids`` or that match one of the ``interface_filters``.
 message InclusiveFilters {
 
-  // A collection of templates.
+  // A collection of templates for which the payload will be included in the
+  // ``create_arguments`` of a matching ``CreatedEvent``.
   // SHOULD NOT contain duplicates.
-  // Required
+  // Optional
   repeated Identifier template_ids = 1;
+
+  // Include an ``InterfaceView`` for every ``InterfaceFilter`` matching a contract.
+  // The ``InterfaceFilter``s MUST use unique ``interface_id``s.
+  // Optional
+  repeated InterfaceFilter interface_filters = 2;
```

Comment:
currently we can have a template_ids subscription referencing future template_ids, which is allowed, and which also correctly starts emitting entries related to the referenced template, as they would appear in the future.

as seen from the respective design document, we do not want to allow unknown interface id-s, as retroactive interfaces make it impossible to have an ongoing stream with retroactive interfaces introduced in the future.

not sure whether users are/will having logic around this aspects, but I think it would be good to document this explicitly. WDYT @meiersi-da @cocreature ?

also: this makes usage different for these two, which would be better to improve. 

one way out would be to leave this difference, allow template_id-s as InterfaceFilter-s (probably with a different name), and for InterfaceFilter-s have consistent handling of not allowing unknown templates and interfaces.
This would solve the aformentioned issue with the include_create_arguments_blob, and would mean that the old filters could be deprecated (InterfaceFilter would be enough in this case)

Reply:
> as seen from the respective design document, we do not want to allow unknown interface id-s, as retroactive interfaces make it impossible to have an ongoing stream with retroactive interfaces introduced in the future.

Good point, I hadn’t considered that. That’s a very good argument to reject those so I retract my comment on the other comment thread.

I’d be inclined to reject unknown template ids as well. In my experience, I’ve only seen users get confused by the empty response instead of a rejection. But would be useful to confirm with product as its technical a breaking change. @stefanobaghino-da wdyt?

Reply:
Let's double check with @stefanobaghino-da whether we want to break the API in this sense for robustness.

Reply:
@nmarton-da I’ve just discussed it with @stefanobaghino-da and he is happy to error out on invalid (unknown or not template) template ids so we can do that consistently everywhere.

Reply:
FYI @remyhaemmerle-da , the map of the interface to templateIds is not enough, we need the set of defined interfaces, and the set of defined templates too

Reply:
note from discussion with @meiersi-da : it is okay to reject with unknown template_id, or interface, because on error applications anyway need to retry, so too eager subscriptions will simply have to retry until the package makes it to the participant.

Reply:
> as seen from the respective design document, we do not want to allow unknown interface id-s, as retroactive interfaces make it impossible to have an ongoing stream with retroactive interfaces introduced in the future.

Forbidding unknown interfaces doesn't solve the problem with retroactive interfaces. The participant may know both the template and the interface ID at the start of the subscription. The retroactive interface declaration can happen in another package uploaded later, can't it?

Reply:
> The retroactive interface declaration can happen in another package uploaded later, can't it?

No it cannot. Retroactive interface declarations must be defined as part of the interface. The analogy here in Haskell are non-orphan instances: The instance must be defined either together with the template or together with the interface. So if you know both, you must also know if there is an instance.

Reply:
Thanks for the clarification.